### ISSUE TITLE
Perfect the player movement animation

### ISSUE THREAD

**Issue by ASteinheiser** [ASteinheiser/react-rpg.com#9]
I've heard multiple times that people feel like the player animation for movement is off. Maybe the transitions speeds need tweaking, or maybe it's the player input debounce rate.

**Comment by ASteinheiser** [comment]
Someone gave me this example: https://github.com/gregoriB/simple-game-React/blob/master/src/components/Player.js
This could help with how the player movement is debounced

**Comment by ASteinheiser** [comment]
b4f71a39dd437344030f0b4b1a8fac206c07e684
2a5a438fa1237d25b9c81c27c814d92cab4abe63
d579248b5a4f403564dd9a9b185f14013c3413bc
1e733e6149c575b78b5b35260aeb976ed83a2787

**Comment by ASteinheiser** [comment]
Even after some tweaking, still doesn't feel right. Can anyone shed some light on what would help make the player movement feel better?

**Comment by Alphanumber** [comment]
I think consulting an animator would be helpful, especially one with experience with pixel art.

In my opinion, the current frames of animation work well enough for continuous movement and would probably also work well enough for moving from cell to cell. However, there are a few quirks with how the animations are being managed that I think make the movement feel jarring and unresponsive.

Arriving in a cell often leaves the character in the middle of a stride. The animation should end on either a frame looking in movement direction or an idle pose frame. I think that when moving from one cell to another, the animation sequence for the movement should only cover one single stride from one down extreme keyframe to the next down keyframe. After finishing a movement, the character should either freeze on the down keyframe or transition to an idle pose/animation. Also, I think the time to transition to a cell could be cut down a fraction.

For more by the books animation additions, there could also be some very simple anticipatory and follow-through action. For anticipatory action, there could be a small windup in the opposite direction of the movement command. If the players enters a move to the right, the character could be bumped to the left roughly  .03 to .04 second before actually initiating the movement to the right. Follow-through action can somewhat be faked where the character could actually overshoot the destination very slightly before settling in finally.


I think most of these additions to the animation can be done in code, but ideally, handcrafted animations would be authored by an animator to be consumed in the game.

I'd suggest looking at references about animation principals to get some basic understanding of some of the history and lessons learned from animators in the past. Richard Williams's _The Animator's Survival Kit_ is highly referenced, but you could also find some tutorials on YouTube.


This YouTube tutorial [here](https://www.youtube.com/watch?v=uDqjIdI4bF4) is a pretty quick rundown on all the principals.

Along with a reference, experimentation and practical application of these principles on animation can lead to insight on how to hack similar results without necessarily rendering them by hand.

**Comment by ASteinheiser** [comment]
Nice video! I think I watched parts of it a while ago, will have to rewatch it.

I like your idea of delaying the animation slightly. And I think you're definitely right that the way the animations are managed is causing some issues. Honestly, I don't know anything about animation, simply tried to implement it with React. All the animations could use a nice refactor.

- Instead of the animations playing then stopping at a random frame, it should play cycles of the animation and know how many cycles to play, or something like that.
- Cycles could be shorter, such as player moving to next cell, which might also make it seem smoother since it's happening faster.

Since you said an actual animator would be best instead doing it in the code, what kind of files would the game consume from the animator? Is it like a LottieFile?

**Comment by Alphanumber** [comment]
I primarily use 3d animation tools so I can't give too much insight on animation for this project. I think largely in the past, most projects designed their own animation format and tools for animators to author animations, especially 2d projects. While Autodesk rules the roost in 3d animation, 2d animation formats are still very much closed and specialized per project, though much less as it has in the past I suspect.

Choosing a third party app we endorse for the project might be a good idea. On the engineering side, we of course want to do as little as possible to be able to consume animation. If there's an animation tool out there with an importer for their proprietary format, then that would be ideal. Trying to reverse engineer a format is not pleasant. 

On the artist side, they'll mainly care about cost and ease-of-use of the content creation and authoring tool. Considering the wants and needs of the artist can help smooth out their processes. I know applications like Spine are popular for use in game development. I've heard a little bit about an online tool called 2dimensions that provide a few runtimes. I don't know anything about LottieFile. Whatever application is considered, the animator should be able to create and export their animations in a format that can be processed by the game.

**Comment by ASteinheiser** [comment]
That's all very true, we have to consider the "artist experience" if we want people to contribute. I certainly want it to be a streamlined as possible.

My knowledge of LottieFiles is that usually they are JS or JSON format and can be exported from Adobe AfterEffects. There seems to be a good library, react-lottie, and I'm sure we can play 1 loop at a time for a smooth animation, but I'm left wondering if this is just another solution that will have the same issues that we have now.

I'll have to do some more research and see what other tools are out there. If you come across anything please share!

**Comment by Alaricus** [comment]
Since the player (and the monsters) only use from their own tile to an adjacent tile each turn, it would make sense to play the entire animation between the starting point and the stopping point. This may require adjusting the length of the animation, but in our case it can be done fairly easily by removing some frames from the sprite-sheet.

For example, frame 0 is currently set to the player standing. We can also set the last from to that, and remove frames 1, 3, 5, and 7. The animation will hardly change (because those are the 'tween' frames) but it will no longer get stuck in mid-motion. The player will simply stand still upon arriving at the destination.

**Comment by ASteinheiser** [comment]
Interesting points @Alaricus. I'm not the best at animation, so I'd definitely like to see how that would pan out. Maybe you could whip up a quick demo?

And yes, as far as I know, all the images and sounds are free for use. The game music was royalty free on youtube, and the sounds were found on a free game asset site. I got the all the tiles and monsters from a free RPG asset site as well. I paid like $4 for the sword slash animations.

That being said, I can't completely remember where I got everything from, so there's a chance I missed something on one image. Should we put a message on the README to contact if you want an image or sound removed?

**Comment by Alaricus** [comment]
I dig deeper into this and I think I figured out what the issue is.

`animationPlayState: animationPlay` is what's causing the seemingly random disconnect between the way the sprite animation should work with `steps()` and the way it actually works. Since `animationPlay` is being set by `setState()` it has no synchronization with what CSS is doing. Sometimes it can fire after a full loop, but the CSS doesn't know when to stop (because it's set to `infinite`) and just keep looping. This results in what looks to us like random pausing points.

I don't know if there is a way to synchronize the firing of `setState()` with CSS animations, but I doubt there is. We are going to need to find a different solution. What immediately comes to mind is a `<canvas>`. We can treat it as just another DOM element for the purposes of movement, but for the purposes of animation we can directly control it with JS.

**Comment by ASteinheiser** [comment]
@Alaricus That makes a lot of sense. To be honest, I just went with the first way that worked for animating the player.

If anyone wants to submit a PR for using canvas based animations, I'd be happy to review.

**Comment by Alaricus** [comment]
I fixed this, will do a PR in a few. However, I just wanted to make sure you are aware that a bunch of tests are currently failing on master. So when you review my PR, please know that I had nothing to do with that. =)

**Comment by ASteinheiser** [comment]
@Alaricus Yeah, I know about that. I need to fix that ASAP. I broke some tests during the big refactor, but didn't get around to fixing those tests.

I'm excited to see your changes though! I hope it's buttery smooth :)

**Comment by ASteinheiser** [comment]
@Alaricus You can see your changes on staging.react-rpg.com now. I think it looks awesome man! I'm going to play around with it a little and then I'll submit a PR for master.

**Comment by ASteinheiser** [comment]
@Alaricus So, after review I think it's almost perfect. Is there any way we could alternate between the "side" the player walks with. For example, when you walk straight down, you will notice the player always puts the same foot and arm forward. It would be nice to take advantage of the other animation variant and alternate when walking. Not sure how easy this would be to accomplish, but this was my only thought.

**Comment by ASteinheiser** [comment]
Closing this issue. We will discuss an animation system and the other game animations in a separate issue.

**Comment by ASteinheiser** [comment]
@Alaricus I noticed a few small things about the player animation that I would like to tackle eventually. Not necessarily you, however I wanted to put this out there for everyone and let you know as well.

I think the "swipe and hold" and "press and hold" to move actions need to have a slower interval; to match the single "key press" debounce rate. I feel like the animation is getting started before the last one can end.

Also, another small thing I noticed is that the player doesn't always alternate sides when stepping. If pressed to quickly, the player will go "left left", "right right". I think we may just need to set the side as the animation starts instead of at the end, or something like that.
